[Product Multi-selection] Resolve toggle state for product multi-selection

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8920

## Description:

This PR addresses the toggle state (selected, and unselected) for `Product` and `Product variations` when product multi-selection is enabled. 

At the moment it is expected that unselecting a product **does not** remove the product from the order, this will be dealt with on a future PR. Both selecting, or unselecting, currently adds a product to the Order.

## Changes:
- `selectProduct` and `selectVariation` are the methods that are called on each view model when we tap on a Product or a Product Variation. We've duplicated the existing `toggleSelection(_:)` logic to account for the closure being called when multi-selection is enabled.
- Renamed `toggleSelectionForVariations` to `toggleSelectionForAllVariations` to reflect that tapping on that toggle will select all variations.
- Removed the call for `toggleSelectionForAllVariations` from the `ProductSelectorView`, as we do not want to select all product variations when the button is tapped.
- Adapted existing Unit Tests to account for the feature flag.
- Added additional Unit Tests to confirm the `productRow` selected state (`.selected` vs `.unselected`) is correct when the feature flag is enabled.

## Testing instructions
- Go to Orders > `+` >  `+ Add Product`
- Add Products to the order. See how the toggle/checkbox is enabled/disabled when tapping
- Go to a variable product and tap on the row. See how you can also select/unselect variable products
- Go back to the Order. Tap on Done.
- Confirm that all products are part of the Order.

| Variable products selection  | Products selection  | Order creation  |
|---|---|---|
| <img width=450 src="https://user-images.githubusercontent.com/3812076/220635012-2e917b31-678b-4743-be91-feef2ced3537.png"> | <img width=450 src="https://user-images.githubusercontent.com/3812076/220635052-c5d7a276-1b8d-44f7-8fe8-37a3992cd1f9.png"> | <img width=450 src="https://user-images.githubusercontent.com/3812076/220635107-76208151-8031-46e3-bf7e-db94150de7cb.png"> |

iamgabrielma committed:
Add toggleSelection to product multi-selection

iamgabrielma committed:
Rename toggleSelectionForVariations, as toggles all variations

iamgabrielma committed:
Add toggleSelection to product variations

iamgabrielma committed:
Remove ability to select all variations in one tap

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -172,16 +172,35 @@ final class ProductSelectorViewModel: ObservableObject {
-        if let onProductSelected = onProductSelected {
-            onProductSelected(selectedProduct)
+
+        if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
I believe you should move this check to Order creation code and add a config for the `ProductSelectorViewModel` instead since this view is reused in different projects.

iamgabrielma replied:
I see! Thanks! I moved it to the `EditableOrderViewModel` instead, and injected the flag into the `ProductSelectorViewModel` on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/12a1d4ba6490a925c4b857598a3e9338f47fb6a7

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorViewModel.swift:
```diff
@@ -159,10 +159,19 @@ final class ProductVariationSelectorViewModel: ObservableObject {
 
-        if let onVariationSelected = onVariationSelected {
-            onVariationSelected(selectedVariation, parentProduct)
+        if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
Same comment regarding adding config instead of checking feature flag directly on the shared view.

iamgabrielma replied:
Changed! https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/429403c061684afce086b72cdd514c3004274728

itsmeichigo reviewed (changes requested):
👋  This works as expected for Order creation, but the product and variation selection no longer works for Coupons:

https://user-images.githubusercontent.com/5533851/220900873-6bb3a787-2dca-4356-9743-3158af7b97fd.mp4

iamgabrielma committed:
Inject feature flag in ProductSelectorViewModel init

iamgabrielma committed:
Extract featureflag from internals of Product selection

iamgabrielma committed:
Rename multi-selection viewmodel flag

iamgabrielma commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -172,16 +181,35 @@ final class ProductSelectorViewModel: ObservableObject {
         configureProductSearch()
     }
 
+    private func debug_helper(product: Product) {
+        print("🍍ProductID: \(product.productID) selected. \(product.name)")
+        print("🍍initialSelectedItems \(initialSelectedItems)")
+        print("🍍selectedProductIDs \(selectedProductIDs)")
+        print("🍍selectedProductVariationIDs \(selectedProductVariationIDs)")
+    }
```
I'll be sure to remove the debug logs before merging :D

iamgabrielma replied:
Removed on 16caac0cbcae3682f22d2ab06526d0165959486a

iamgabrielma commented:
Thanks for the review! I extracted the feature flag from the Product and ProductVariation view models, so the previous Coupons logic are not affected by the changes. This is ready for re-review when you have the time :)

iamgabrielma committed:
Remove debug helper method

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -178,10 +187,20 @@ final class ProductSelectorViewModel: ObservableObject {
-            onProductSelected(selectedProduct)
+
+        if shouldToggleItemOnSelection {
+            // Product multi-selection (new)
+            if let onProductSelected = onProductSelected {
+                toggleSelection(productID: productID)
+                onProductSelected(selectedProduct)
+            }
         } else {
-            toggleSelection(productID: productID)
+            // Product single-selection (legacy)
+            if let onProductSelected = onProductSelected {
+                onProductSelected(selectedProduct)
+            } else {
+                toggleSelection(productID: productID)
+            }
```
💡 This is more of a nit: I'd prefer not calling the new logic multi-selection, since we have multi-selection before and it just behaves in a slightly different way. How about structuring these checks in another way to make it clearer to see the differences:

```Swift
        if shouldToggleItemOnSelection, let onProductSelected {
            // toggles the item and then triggers the selection
            toggleSelection(productID: productID)
            onProductSelected(selectedProduct)
        } else if let onProductSelected {
            // the selector supports single selection only, so the closure should be triggered and the work is done.
            onProductSelected(selectedProduct)
        } else {
            // no closure for single item selection is available; toggles the selection only for multiple selections.
            toggleSelection(productID: productID)
        }
```

itsmeichigo replied:
Looking at the suggested change above, I think maybe renaming `shouldToggleItemOnSelection` into `supportsMultipleSelections` would be more precise. WDYT?

iamgabrielma replied:
> How about structuring these checks in another way to make it clearer to see the differences:

Looks great, and less verbose than my approach, changed on 976e6e2f4d8e59055c7c34f7ef1dc49ce1d90f50

> renaming

I can't seem to figure out a proper name for this flag, as should not only convey multi-selection is allowed but also that there's logic happening after each selection (unlike coupons). That said, `supportsMultipleSelections` definitely sounds better than `shouldToggleItemOnSelection`, I can re-think variable this once we move to M1.1, as the logic changes slightly as well. Changed on e8bdc191ff7d4dcb20439fa12758ca1bc3c1508a

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorViewModel.swift:
```diff
@@ -159,10 +168,19 @@ final class ProductVariationSelectorViewModel: ObservableObject {
-        if let onVariationSelected = onVariationSelected {
-            onVariationSelected(selectedVariation, parentProduct)
+        if shouldToggleItemOnSelection {
+            // Product variation multi-selection (new)
+            if let onVariationSelected = onVariationSelected {
+                toggleSelection(productVariationID: variationID)
+                onVariationSelected(selectedVariation, parentProduct)
+            }
         } else {
-            toggleSelection(productVariationID: variationID)
+            // Product variation single-selection (legacy)
+            if let onVariationSelected = onVariationSelected {
+                onVariationSelected(selectedVariation, parentProduct)
+            } else {
+                toggleSelection(productVariationID: variationID)
+            }
```
💡  Same comment as above for readability.

iamgabrielma replied:
Changed on f06999092d5225f5a9a72a2f2c2cf1211ff286fd

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -345,7 +345,28 @@ final class ProductSelectorViewModelTests: XCTestCase {
         let productRow = viewModel.productRows.first(where: { $0.productOrVariationID == product.productID })
         XCTAssertNotNil(productRow)
-        XCTAssertEqual(productRow?.selectedState, .selected)
+        if !ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
💡 For tests, it's recommended to use `MockFeatureFlagService` and inject it into the view models. That way, you can set the value for any flag and the tests will not fail if the actual flag is updated. You will not have to put the assertions in a conditional check either.

However, in this case we are testing `ProductSelectorViewModel`, which has been updated with the new flag, so you should inject the value into the initializer to test this case.

iamgabrielma replied:
> MockFeatureFlagService

Thanks for pointing that out, I didn't know about `MockFeatureFlagService` 🙇 

Injected the flag for all tests into the initializer on 82dfbb86c6d893a8526ce449998d7b438dc5c5c4 , rather than relying on the feature flag check.

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -345,7 +345,28 @@ final class ProductSelectorViewModelTests: XCTestCase {
+        viewModel.selectProduct(product.productID)
+
+        // Then
+        if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
Same comment as above about remove feature flag service check here.

iamgabrielma replied:
Changed: https://github.com/woocommerce/woocommerce-ios/commit/82dfbb86c6d893a8526ce449998d7b438dc5c5c4

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -365,6 +386,26 @@ final class ProductSelectorViewModelTests: XCTestCase {
+        viewModel.selectProduct(product.productID)
+
+        // Then
+        if ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
Same comment as above about removing the check with feature flag service.

iamgabrielma replied:
Changed: https://github.com/woocommerce/woocommerce-ios/commit/82dfbb86c6d893a8526ce449998d7b438dc5c5c4

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -470,7 +511,9 @@ final class ProductSelectorViewModelTests: XCTestCase {
 
         // Then
-        XCTAssertEqual(selectedItems, [simpleProduct.productID, 12])
+        if !ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
Please also replace this check and inject value for `shouldToggleItemOnSelection` in the initializer instead.

iamgabrielma replied:
Changed: https://github.com/woocommerce/woocommerce-ios/commit/82dfbb86c6d893a8526ce449998d7b438dc5c5c4

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -530,12 +573,16 @@ final class ProductSelectorViewModelTests: XCTestCase {
         // Confidence check
         let productRow = viewModel.productRows.first(where: { $0.productOrVariationID == product.productID })
-        XCTAssertEqual(productRow?.selectedState, .selected)
+        if !ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
Same comment as above.

iamgabrielma replied:
Changed: https://github.com/woocommerce/woocommerce-ios/commit/82dfbb86c6d893a8526ce449998d7b438dc5c5c4

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Orders/Order Creation/ProductSelectorViewModelTests.swift:
```diff
@@ -530,12 +573,16 @@ final class ProductSelectorViewModelTests: XCTestCase {
         // Then
         let updatedRow = viewModel.productRows.first(where: { $0.productOrVariationID == product.productID })
-        XCTAssertEqual(updatedRow?.selectedState, .notSelected)
+        if !ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1) {
```
Same comment as above.

iamgabrielma replied:
Changed: https://github.com/woocommerce/woocommerce-ios/commit/82dfbb86c6d893a8526ce449998d7b438dc5c5c4

itsmeichigo reviewed (commented):
Thanks for the updates! This looks better, but I think can still be improved in a few places 😅

iamgabrielma committed:
Tests: remove feature flag in favor of injecting flag on init

iamgabrielma committed:
Refactor toggling single vs multiple selection

iamgabrielma committed:
Refactor toggling single vs multiple selection for variations

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -52,6 +52,11 @@ final class ProductSelectorViewModel: ObservableObject {
     @Published private(set) var productRows: [ProductRowViewModel] = []
 
+    /// Determines if products should be toggled on selection
+    /// `false` by default, mainly used to allow multi-seleciton on the Order Creation flow
```
typo here multi-seleciton

iamgabrielma replied:
Fixed! a7ddbece2a2ef1aca3708919acdd871b300b7a62

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -52,6 +52,11 @@ final class ProductSelectorViewModel: ObservableObject {
+    /// Determines if products should be toggled on selection
+    /// `false` by default, mainly used to allow multi-seleciton on the Order Creation flow
+    ///
+    let shouldToggleItemOnSelection: Bool
```
I am trying to understand it here 😅 What does it mean toggle on selection, and why would we want that they are not toggled?

iamgabrielma replied:
> I am trying to understand it here 😅 What does it mean toggle on selection, and why would we want that they are not toggled?

That's a good question, I'm trying to come up with a proper name for this flag, for the moment has been changed to `supportsMultipleSelections` in the latest commits ([Convo here](https://github.com/woocommerce/woocommerce-ios/pull/8935#discussion_r1117872002)). 

The flag is used to determine if we allow items to be multi-selected, and logic happens when we select them. This is the main difference between the Order and the Coupons implementation, the latter allows product multi-selection but nothing is triggered while products are selected (just UI), while in the Order implementation we trigger a callback each time a product is selected, so we can add/remove them to the Order.

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -178,10 +187,20 @@ final class ProductSelectorViewModel: ObservableObject {
+            }
         } else {
-            toggleSelection(productID: productID)
+            // Product single-selection (legacy)
```
Is this legacy being used elsewhere, or can we remove it completely?

iamgabrielma replied:
> Is this legacy being used elsewhere, or can we remove it completely?

This part has been refactored here: https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/976e6e2f4d8e59055c7c34f7ef1dc49ce1d90f50 for easier readability

We currently need it for the case where we use the Product Selector on single item selections, as this will toggle the item without any callback.

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorViewModel.swift:
```diff
@@ -16,6 +16,11 @@ final class ProductVariationSelectorViewModel: ObservableObject {
+    /// Determines if product variations should be toggled on selection
+    /// `false` by default, mainly used to allow multi-seleciton on the Order Creation flow
+    ///
+    let shouldToggleItemOnSelection: Bool
```
It seems like we are duplicating logic here, can we somehow have or extract the logic to a common class for both?

iamgabrielma replied:
> It seems like we are duplicating logic here, can we somehow have or extract the logic to a common class for both?

I'd prefer to not change this part within the scope of this PR, as the behaviour might be different for the product selector and the product variation selector, specially since is also used with Coupons. 

Happy to revisit it outside of this PR though, for example I think we could pass some sort of common Configuration class to the initializer, or use generics to abstract some bits of the logic on both view models. Opened a new issue to keep track of this : https://github.com/woocommerce/woocommerce-ios/issues/8988

iamgabrielma committed:
typo

iamgabrielma committed:
Rename supportsMultipleSelections

iamgabrielma committed:
Add & update Unit Tests with supportsMultiSelections

iamgabrielma committed:
Add button title configuration to multiselection

iamgabrielma committed:
Whitespace

iamgabrielma commented:
Thanks for the review @toupper and @itsmeichigo ! This is ready for another check when possible 🙇

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -151,9 +151,7 @@ struct ProductSelectorView: View {
-                           viewModel: rowViewModel) {
-                    viewModel.toggleSelectionForVariations(of: rowViewModel.productOrVariationID)
-                }
+                           viewModel: rowViewModel)
```
I missed this change in the last review 😅 I think for coupons we want to be able to select all variations for a product (to include or exclude them from a coupon) - so this callback should not be removed.

We should add another config for the `ProductSelectorViewModel` to indicate whether this is the behavior or not with something like `enableTogglingSelectionForAllVariations`.

iamgabrielma replied:
I see what you mean, somehow I missed this behaviour too. I made the changes here: https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/bc6e1c49004c03774fe985b277d5ef3c7c3ecfa9 , leaving the toggle for all variations to true by default and disallowing it in the ProductSelector.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift:
```diff
@@ -168,13 +168,18 @@ final class EditableOrderViewModel: ObservableObject {
+            purchasableItemsOnly: true,
+            storageManager: storageManager,
+            stores: stores,
+            supportsMultipleSelections: ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1)) { [weak self] product in
```
Nit: For initializers that include more than one closure, it's best to keep their names rather than using trailing closures to ensure readability:

```suggestion
            supportsMultipleSelections: ServiceLocator.featureFlagService.isFeatureFlagEnabled(.productMultiSelectionM1), onProductSelected: { [weak self] product in
```

iamgabrielma replied:
Totally agree. I changed this on other places, but I may have missed this one. Changed on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/79598230157639942abf6482f6baf25d742ca670

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorViewModel.swift:
```diff
@@ -16,6 +16,11 @@ final class ProductVariationSelectorViewModel: ObservableObject {
     private let stores: StoresManager
 
+    /// Determines if multiple item selection is supported, `false` by default.
+    /// Mainly used to allow multi-selection on the Order Creation flow
```
We should not include this comment as this is too specific for a shared view, which makes it sound like it's not applicable to other features 😅

iamgabrielma replied:
Removed! https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/a52c1b23598c20a0ebe6fba77b651cc984b4b916

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -178,7 +187,13 @@ final class ProductSelectorViewModel: ObservableObject {
+        } else if let onProductSelected {
+            // The selector supports single selection only
             onProductSelected(selectedProduct)
         } else {
             toggleSelection(productID: productID)
```
I also missed the fact that coupons support multiple selections as well, but now the feature still work without setting `supportsMultipleSelections` to true for those selectors. This check is not reliable enough then, we should add a check for when `supportsMultipleSelections` is true too:

```suggestion
        } else if supportsMultipleSelections {
            toggleSelection(productID: productID)
```

Then when `supportsMultipleSelections` is false, no action is taken here.

That would require changing the initializers of the product selectors in `AddEditCouponViewModel` and `CouponRestrictionsViewModel` to set `supportsMultipleSelections` to `true` - would you mind updating that too? Thanks!

iamgabrielma replied:
Oh I see what you mean, I updated the view models here: https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/58074731c6fb4314a58dbacc5037efd7d910c0bd and removed a unit test that will always fail, as we're not checking for row state on single selection anymore when toggling an item.

itsmeichigo reviewed (changes requested):
Thanks for the changes! I found the some issues that I missed in the last review, please check out the comments below.

iamgabrielma committed:
Remove trailing closure for readablity

iamgabrielma committed:
Internal comment

iamgabrielma committed:
Toggling all variations on products vs coupons

iamgabrielma committed:
Update Coupons to new multiple selection support

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -52,6 +52,15 @@ final class ProductSelectorViewModel: ObservableObject {
     @Published private(set) var productRows: [ProductRowViewModel] = []
 
+    /// Determines if multiple item selection is supported, `false` by default.
+    /// Mainly used to allow multi-selection on the Order Creation flow
```
Not a blocker, but this comment feels to specific for a generic class like this one

iamgabrielma replied:
Ah, yes! Removed on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/03b231d9239b592c5846d442c2a9e3495b4532c0

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -52,6 +52,15 @@ final class ProductSelectorViewModel: ObservableObject {
+    ///
+    let supportsMultipleSelections: Bool
+
+    /// Determines if is possible to toggle all variatoon items upon selection
```
Typos in comment (if **it** is possible ... all **variation**)

iamgabrielma replied:
Fixed typo on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/03b231d9239b592c5846d442c2a9e3495b4532c0

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -52,6 +52,15 @@ final class ProductSelectorViewModel: ObservableObject {
+    /// Determines if multiple item selection is supported, `false` by default.
+    /// Mainly used to allow multi-selection on the Order Creation flow
+    ///
+    let supportsMultipleSelections: Bool
```
Not a blocker, I would make it singular (`supportsMultipleSelection`). To make it more readable, I would consider using an enum for this:

```
enum SelectionStyle {
case simple
case multiple
}
```

iamgabrielma replied:
Thanks! Made the property singular on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/65fe4ae1998a785ccdcf879905bb5dda915171a3 . I'll leave the usage of an enum for https://github.com/woocommerce/woocommerce-ios/issues/8988 when trying to abstract properties for both Products and Coupons, I believe we may get rid of this bit entirely though, as by default all selections would be multi-selections.

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -52,6 +52,15 @@ final class ProductSelectorViewModel: ObservableObject {
+
+    /// Determines if is possible to toggle all variatoon items upon selection
+    ///
+    let enableTogglingSelectionForAllVariations: Bool
```
I think the comment pattern is more meaningful for this property naming:

`let toggleAllVariationsOnSelection`

iamgabrielma replied:
> comment pattern

Interesting! Definitely more readable, changed on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/e1142d11405dbc38ed7317ea85d6538e5842c088

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -132,11 +141,15 @@ final class ProductSelectorViewModel: ObservableObject {
          purchasableItemsOnly: Bool = false,
          storageManager: StorageManagerType = ServiceLocator.storageManager,
          stores: StoresManager = ServiceLocator.stores,
+         supportsMultipleSelections: Bool = false,
```
Why do we set these default values?

iamgabrielma replied:
Currently we set these to `false` on initialization because is the default way to select products (single-selection), then we either pass `true` on initialization when creating the viewmodels for Coupons, or via if needed via feature flag for Product multi-selection.

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -178,9 +195,15 @@ final class ProductSelectorViewModel: ObservableObject {
         }
-        if let onProductSelected = onProductSelected {
+
+        if supportsMultipleSelections, let onProductSelected {
```
I think we can simplify these if-else blocks, I let you find a more optimal and shorter way to do it :)

iamgabrielma replied:
Right! I refactored these using `guard` on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/47c84cac6403d261a1c6818807681adc06f0abde . Works as expected and all test pass, but happy to get a second look to the changes, as personally I still don't find them more readable using guard rather than "if else", I have to get better acquainted :P

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorViewModel.swift:
```diff
@@ -16,6 +16,10 @@ final class ProductVariationSelectorViewModel: ObservableObject {
 
+    /// Determines if multiple item selection is supported
+    ///
+    let supportsMultipleSelections: Bool
```
same here about naming

iamgabrielma replied:
Changed as well with https://github.com/woocommerce/woocommerce-ios/commit/65fe4ae1998a785ccdcf879905bb5dda915171a3

toupper commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductVariationSelectorViewModel.swift:
```diff
@@ -159,7 +167,12 @@ final class ProductVariationSelectorViewModel: ObservableObject {
         }
 
-        if let onVariationSelected = onVariationSelected {
+        if supportsMultipleSelections, let onVariationSelected {
```
same here about if-else blocks

iamgabrielma replied:
Changed as well with https://github.com/woocommerce/woocommerce-ios/commit/47c84cac6403d261a1c6818807681adc06f0abde

iamgabrielma committed:
Internal comment. Typo

iamgabrielma committed:
Rename variations toggle property for clarity

iamgabrielma committed:
Rename property from plural to singular

iamgabrielma committed:
Refactor selectProduct to use guard

iamgabrielma commented:
This is ready for re-review after the latest changes @toupper 🙇

iamgabrielma committed:
Merge pull request #8987 from woocommerce/issue/8938-multiselection-button-configuration

[Product Multi-selection] Adapt "done" button configuration for multiple products selection

iamgabrielma committed:
Turn feature flag off

toupper commented:
@iamgabrielma  The code looks good to me now, but when testing I noticed that the UI isn't blocked when adding a new element in the background, which might be confusing. Furthermore, as seen in your screenshot above, the purple "Done" button below loses its title randomly when selecting products.

iamgabrielma committed:
Disallow selecting a product if it’s already synching

By adding a check to `shouldShowScrollIndicator` when selecting a Product, we’re blocking the merchant from tapping multiple times while the Order is synching, effectively blocking the UI.

iamgabrielma commented:
Thanks again for the review! I believe we should be there with the latest changes! 🍿 

> UI isn't blocked when adding a new element in the background

I addressed this on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/2bd9cb27355fd6b4b044fc2ff00d2fefdf3d409e by returning early from the `selectProduct(_:)` method if Products are still synchronising, so nothing happens if the merchant taps again any of the products if the sync to remote is not complete. I tried different approaches, like blocking the UI by showing a loading screen, but this approach seemed the cleanest with the current implementation.

> the purple "Done" button below loses its title randomly when selecting products

I could have done a better job splitting PRs on a different manner, so it would have been less confusing to see what's done and what's not, sorry for that 😅 This was resolved on a [different PR](https://github.com/woocommerce/woocommerce-ios/pull/8987) that I just merged into this branch, so you should see the `Done` button working fine at the moment.

Please note that at the moment it is expected that both selecting and unselecting a Product adds such Product to an Order (unselecting does not remove them from the Order), this will be dealt with on a [different PR](https://github.com/woocommerce/woocommerce-ios/pull/8943)

toupper reviewed (approved):
@iamgabrielma Thanks for clarifying! It looks good now, looking forward to reviewing other PRs

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorView.swift:
```diff
@@ -151,9 +151,12 @@ struct ProductSelectorView: View {
+                           onCheckboxSelected: {
+                    if viewModel.toggleAllVariationsOnSelection {
+                        viewModel.toggleSelectionForAllVariations(of: rowViewModel.productOrVariationID)
+                    }
```
Nit: You can check for `toggleAllVariationsOnSelection` within the implementation of `toggleSelectionForAllVariations` method to simplify the logic in the view.

iamgabrielma replied:
Good catch, thanks! I moved the logic to the view model on https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/b6757b0aa19a45747b474c16dbcf789710d39877

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Products/ProductSelector/ProductSelectorViewModel.swift:
```diff
@@ -175,14 +191,26 @@ final class ProductSelectorViewModel: ObservableObject {
     /// Select a product to add to the order
     ///
     func selectProduct(_ productID: Int64) {
+        // If it's synching and there's an attempt to select a product, return early
+        if shouldShowScrollIndicator {
+            return
+        }
```
I don't think this property `shouldShowScrollIndicator` has anything to do with product selection. Commenting out this code works the same way. I'd suggest removing it.

iamgabrielma replied:
Thanks for digging into this. I see we're calling it through sync, but not when we do product selection but whether the product list is loading more content. I removed it here https://github.com/woocommerce/woocommerce-ios/pull/8935/commits/25b7bdfab3c4142ffa2d0d9f17a33f681049cdeb , and confirmed that we still block the UI when the product is being selected, disallowing from selecting another one ( which is what we want ). 

Something that can be confusing when testing this and checking the Order after selecting a bunch of products is what I mentioned [here](https://github.com/woocommerce/woocommerce-ios/pull/8935#issuecomment-1449881698) regarding both selecting and unselecting a Product adds such Product to an Order (unselecting does not remove them from the Order), this will be dealt with on a https://github.com/woocommerce/woocommerce-ios/pull/8943 and I can test further if there's any issue blocking the UI, or revisit how we reflect this to the end-user.

iamgabrielma committed:
remove shouldShowScrollIndicator check for selectProduct

iamgabrielma committed:
Move toggleAllVariationsOnSelection to view model

itsmeichigo reviewed (approved):
Thanks for your patience!
